gateway/src/node/storage/bandwidth.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2021 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::node::storage::models::PersistedBandwidth;
+
+#[derive(Clone)]
+pub(crate) struct BandwidthManager {
+    connection_pool: sqlx::SqlitePool,
+}
+
+impl BandwidthManager {
+    /// Creates new instance of the `BandwidthManager` with the provided sqlite connection pool.
+    ///
+    /// # Arguments
+    ///
+    /// * `connection_pool`: database connection pool to use.
+    pub(crate) fn new(connection_pool: sqlx::SqlitePool) -> Self {
+        BandwidthManager { connection_pool }
+    }
+
+    /// Creates a new bandwidth entry for the particular client.
+    ///
+    /// # Arguments
+    ///
+    /// * `client_address_bs58`: base58-encoded address of the client.
+    pub(crate) async fn insert_new_client(
+        &self,
+        client_address_bs58: &str,
+    ) -> Result<(), sqlx::Error> {
+        sqlx::query!(
+            "INSERT INTO available_bandwidth(client_address_bs58, available) VALUES (?, 0)",
+            client_address_bs58
+        )
+        .execute(&self.connection_pool)
+        .await?;
+        Ok(())
+    }
+
+    /// Tries to retrieve available bandwidth for the particular client.
+    ///
+    /// # Arguments
+    ///
+    /// * `client_address_bs58`: base58-encoded address of the client.
+    pub(crate) async fn get_available_bandwidth(
+        &self,
+        client_address_bs58: &str,
+    ) -> Result<Option<PersistedBandwidth>, sqlx::Error> {
+        sqlx::query_as!(
+            PersistedBandwidth,
+            "SELECT * FROM available_bandwidth WHERE client_address_bs58 = ?",
+            client_address_bs58
+        )
+        .fetch_optional(&self.connection_pool)
+        .await
+    }
+
+    /// Increases available bandwidth of the particular client by the specified value.
+    ///
+    /// # Arguments
+    ///
+    /// * `client_address_bs58`: base58-encoded address of the client.
+    /// * `value`: amount of available bandwidth to be added to the client.
+    pub(crate) async fn increase_available_bandwidth(
+        &self,
+        client_address_bs58: &str,
+        value: i64,
+    ) -> Result<(), sqlx::Error> {
+        sqlx::query!(
+            r#"
+                UPDATE available_bandwidth
+                SET available = available + ?
```

Comment:
How does `sqlx` handle overflow situations?

Reply:
That is a very interesting question so I've run an experiment. Basically `sqlite` will happily store data bigger than maximum size of `int64` (so we never lose any "overflow" bandwidth). However, on query the results from `sqlx` are capped at the maximum value, meaning that in the unlikely scenario of client having more than few exabytes of traffic available, he is not going to see anything above it (but it would still exist in the database)